source/extensions/filters/http/set_metadata/set_metadata_filter.cc

```diff
@@ -0,0 +1,49 @@
+#include "extensions/filters/http/set_metadata/set_metadata_filter.h"
+
+#include "envoy/extensions/filters/http/set_metadata/v3/set_metadata.pb.h"
+
+#include "common/config/well_known_names.h"
+#include "common/http/utility.h"
+#include "common/protobuf/protobuf.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace SetMetadataFilter {
+
+Config::Config(const envoy::extensions::filters::http::set_metadata::v3::Config& proto_config) {
+  namespace_ = proto_config.metadata_namespace();
+  value_ = proto_config.value();
+}
+
+SetMetadataFilter::SetMetadataFilter(const ConfigSharedPtr config) : config_(config) {}
+
+SetMetadataFilter::~SetMetadataFilter() = default;
+
+Http::FilterHeadersStatus SetMetadataFilter::decodeHeaders(Http::RequestHeaderMap&, bool) {
+  const auto metadata_namespace = config_->metadataNamespace();
+  auto& metadata = *decoder_callbacks_->streamInfo().dynamicMetadata().mutable_filter_metadata();
+  ProtobufWkt::Struct& org_fields = metadata[metadata_namespace];
+  ProtobufWkt::Struct const& to_merge = config_->value();
+
+  StructUtil::update(org_fields, to_merge);
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus SetMetadataFilter::decodeData(Buffer::Instance&, bool) {
+  return Http::FilterDataStatus::Continue;
```

Comment:
Nit: don't think this is covered in tests.

Reply:
Added in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c , by calling `decodeData` in `runFilter`